File: LeetCode/CodingInterviews/exchange.go
package coding

func exchange(nums []int) []int {
	var i, j = 0, len(nums) - 1
	for i < j {
		for i < j && nums[i]&1 == 1 {
			i++
		}
		for i < j && nums[j]&1 == 0 {
			j--
		}
		nums[i], nums[j] = nums[j], nums[i]
	}
	return nums
}

File: LeetCode/CodingInterviews/Java/Exchange.java
package LeetCode;

class Solution {
    public int[] exchange(int[] nums) {
        int l = 0, r = nums.length - 1;
        while (l < r) {
            while (nums[l] % 2 == 1 && l < r)
                ++l;
            while (nums[r] % 2 == 0 && l < r)
                --r;
            int t = nums[l];
            nums[l] = nums[r];
            nums[r] = t;
        }
        return nums;
    }
}